sled-agent/src/support_bundle/logs.rs

```diff
@@ -0,0 +1,253 @@
+            };
+
+            info!(&self.log, "SupportBundle GET log file (ranged)";
+                "file" => %valid_path,
+                "start" => range.start(),
+                "limit" => range.content_length().get(),
+            );
+
+            file.seek(std::io::SeekFrom::Start(range.start())).await?;
+            let limit = range.content_length().get();
+            return Ok(range_requests::make_get_response(
+                Some(range),
+                metadata.len(),
+                content_type,
+                ReaderStream::new(file.take(limit)),
+            )?);
+        } else {
+            info!(&self.log, "SupportBundle GET log file";
+                "file" => %valid_path,
+            );
+
+            Ok(range_requests::make_get_response(
+                None,
+                metadata.len(),
+                content_type,
+                // The take here is load bearing as we want to cap continously
+                // growing log files.
+                ReaderStream::new(file.take(metadata.len())),
+            )?)
+        }
+    }
+}
+
+/// Given a path represented as a `str` return a normalized `Utf8PathBuf` that
+/// starts with "/" and removes "..", "."  components.
+fn normalize_path(path: &str) -> Utf8PathBuf {
+    let path = Utf8PathBuf::from(format!("/{path}"));
+    let mut normalized = Utf8PathBuf::new();
+
+    for component in path.components() {
+        match component {
+            camino::Utf8Component::RootDir => normalized.push("/"),
+            camino::Utf8Component::Normal(c) => normalized.push(c),
+            // Toss the rest out rather than popping off the stack. Consumers
+            // should be passing full paths.
+            _ => (),
+        }
+    }
+
+    normalized
+}
+
+fn validate_log_dir(
+    allowed_dirs: &BTreeSet<Utf8PathBuf>,
+    path: Utf8PathBuf,
+) -> Result<Utf8PathBuf, Error> {
+    let directory =
+        path.parent().ok_or(Error::MissingFilename(path.clone()))?;
+    match allowed_dirs.contains(directory) {
+        true => Ok(path.clone()),
+        false => Err(Error::InvalidLogDir(path)),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::support_bundle::logs::*;
+    use camino::Utf8PathBuf;
+
+    #[test]
+    fn test_normalize_path() {
+        assert_eq!(
+            normalize_path("missing/root/dir"),
+            Utf8PathBuf::from(r"/missing/root/dir")
+        );
+
+        assert_eq!(
+            normalize_path("/etc/passwd"),
+            Utf8PathBuf::from(r"/etc/passwd")
+        );
+
+        assert_eq!(
+            normalize_path("../../etc/shadow"),
+            Utf8PathBuf::from(r"/etc/shadow")
+        );
+
+        assert_eq!(
+            normalize_path("./../secret/file"),
+            Utf8PathBuf::from(r"/secret/file")
+        );
+
+        assert_eq!(
+            normalize_path("/find/the/./../secret/file"),
+            Utf8PathBuf::from(r"/find/the/secret/file")
```

Comment:
This seems *wrong* to me though. I understand wanting to not parse `..` here, but claiming this is a "normalized" variant of the path does not seem correct.

I wish there was something in the stdlib for this - `canonicalize` exists, but it tries to walk symlinks, which we don't necessarily need.

But https://github.com/rust-lang/cargo/blob/fede83ccf973457de319ba6fa0e36ead454d2e20/src/cargo/util/paths.rs#L61-L86 is how cargo defines normalization, and it treats `..` as a signal to pop from the path. We **don't need to do that**, but as mentioned earlier, ignoring it seems incorrect.

Reply:
Yea `canonicalize` as far as I remember actually examines the filesystem rather than attempting to normalize the path. As stated above should we just return an error if we encounter a `..` path component?